data_diff/databases/clickhouse.py

```diff
@@ -0,0 +1,153 @@
+from .base import MD5_HEXDIGITS, CHECKSUM_HEXDIGITS
+from .database_types import (
+    ColType, Decimal, Float, Integer, FractionalType, Native_UUID, TemporalType, Text, Timestamp
+)
+
+
+@import_helper("clickhouse")
+def import_clickhouse():
+    import clickhouse_driver
+    return clickhouse_driver
+
+
+class SingleConnection(clickhouse_driver.dbapi.connection.Connection):
+    def cursor(self, cursor_factory=None):
+        if not len(self.cursors):
+            _ = super().cursor()
+        return self.cursors[0]
+
+
+class Clickhouse(ThreadedDatabase):
+    TYPE_CLASSES = {
+        'Int8': Integer,
+        'Int16': Integer,
+        'Int32': Integer,
+        'Int64': Integer,
+        'Int128': Integer,
+        'Int256': Integer,
+
+        'UInt8': Integer,
+        'UInt16': Integer,
+        'UInt32': Integer,
+        'UInt64': Integer,
+        'UInt128': Integer,
+        'UInt256': Integer,
+
+        'Float32': Float,
+        'Float64': Float,
+
+        'Decimal': Decimal,
+
+        'UUID': Native_UUID,
+
+        'String': Text,
+        'FixedString': Text,
+
+        'DateTime': Timestamp,
+        'DateTime64': Timestamp,
+
+    }
+    ROUNDS_ON_PREC_LOSS = True
+
+    def __init__(self, *, thread_count: int, **kw):
+        super().__init__(thread_count=thread_count)
+
+        self._args = kw
+        # In Clickhouse database and schema are the same
+        self.default_schema = kw['database']
+
+    def create_connection(self):
+        clickhouse = import_clickhouse()
+        try:
+            connection = SingleConnection(**self._args)
+            return connection
+        except clickhouse.OperationError as e:
+            raise ConnectError(*e.args) from e
+
+    def _parse_type_repr(self, type_repr: str) -> Optional[Type[ColType]]:
+        nullable_prefix = 'Nullable'
+        if type_repr.lower().startswith(nullable_prefix.lower()):
+            type_repr = type_repr[len(nullable_prefix):].lstrip('(').rstrip(')')
+
+        if type_repr.startswith('Decimal'):
+            type_repr = 'Decimal'
+        elif type_repr.startswith('FixedString'):
+            type_repr = 'FixedString'
+        elif type_repr.startswith('DateTime64'):
+            type_repr = 'DateTime64'
+
+        return self.TYPE_CLASSES.get(type_repr)
+
+    def quote(self, s: str) -> str:
+        return f'"{s}"'
+
+    def md5_to_int(self, s: str) -> str:
+        substr_idx = 1 + MD5_HEXDIGITS - CHECKSUM_HEXDIGITS
+        return f'reinterpretAsUInt128(reverse(unhex(lowerUTF8(substr(hex(MD5({s})), {substr_idx})))))'
+
+    def to_string(self, s: str) -> str:
+        return f"toString({s})"
+
+    def normalize_timestamp(self, value: str, coltype: TemporalType) -> str:
+        if coltype.rounds:
+            prec = coltype.precision
+            timestamp = f'toDateTime64(round(toUnixTimestamp64Micro(toDateTime64({value}, 6)) / 1000000, {prec}), 6)'
+            return self.to_string(timestamp)
+        else:
```

Comment:
"else" clause not needed. I suggest using pylint on this module to catch things like that.

Reply:
Done.